parser2/src/tokenize.rs

```diff
@@ -0,0 +1,585 @@
+
+//                 // New context should not be unindented past the closest enclosing block context
+//                 Context::MatchArm |
+//                 Context::Type |
+//                 Context::Let |
+//                 Context::Block { .. } if offside.column < other_offside.column => {
+//                     return Err(Error::UnderIndented)
+//                 }
+
+//                 Context::MatchArm |
+//                 Context::Type |
+//                 Context::Let |
+//                 Context::Block { .. } |
+//                 Context::Expr |
+//                 Context::If |
+//                 Context::Then => (),
+//             }
+//         }
+
+//         Ok(())
+//     }
+// }
+
+pub struct Tokenizer<'input> {
+    src: &'input str,
+    chars: CharIndices<'input>,
+    lookahead: Option<(usize, char)>,
+    // indent_levels: Contexts,
+}
+
+fn is_operator_symbol(ch: char) -> bool {
+    "+-*/&|=<>:.".chars().any(|x| x == ch)
+}
+
+fn is_ident_start(ch: char) -> bool {
+    UnicodeXID::is_xid_start(ch)
+}
+
+fn is_ident_continue(ch: char) -> bool {
+    UnicodeXID::is_xid_continue(ch)
+}
+
+fn is_digit(ch: char) -> bool {
+    ch.is_digit(10)
+}
+
+fn escape_code(c: char) -> Result<char, Error> {
+    match c {
+        '\'' => Ok('\''),
+        '"' => Ok('"'),
+        '\\' => Ok('\\'),
+        '/' => Ok('/'),
+        'n' => Ok('\n'),
+        'r' => Ok('\r'),
+        't' => Ok('\t'),
+        // TODO: Unicode escape codes
+        ch => Err(Error::UnexpectedEscapeCode(ch)),
+    }
+}
+
+impl<'input> Tokenizer<'input> {
+    pub fn new(src: &'input str) -> Tokenizer<'input> {
+        let mut chars = src.char_indices();
+
+        Tokenizer {
+            src: src,
+            lookahead: chars.next(),
+            chars: chars,
+            // indent_levels: Contexts::new(),
+        }
+    }
+
+    fn bump(&mut self) -> Option<(usize, char)> {
+        match self.lookahead {
+            Some((_byte, _ch)) => {
+                // if ch == '\n' {
+                //     self.location.column = 0;
+                //     self.location.line += 1;
+                // } else {
+                //     // TODO: Handle carriage returns
+                //     self.location.column = 1;
+                // }
+                // self.location.byte += byte;
+                self.lookahead = self.chars.next();
+                self.lookahead
+            }
+            None => None,
+        }
+    }
+
+    fn take_while<F>(&mut self, mut keep_going: F) -> Option<usize>
+        where F: FnMut(char) -> bool
+    {
+        self.take_until(|c| !keep_going(c))
+    }
+
+    fn take_until<F>(&mut self, mut terminate: F) -> Option<usize>
+        where F: FnMut(char) -> bool
+    {
+        while let Some((i1, ch)) = self.lookahead {
+            if terminate(ch) {
+                return Some(i1);
+            } else {
+                self.bump();
+            }
+        }
+        None
+    }
```

Comment:
Is there a need for detecting the EOF case? If not it seems simpler to just return a `usize` with what is currently `None` returned as the length of the string.